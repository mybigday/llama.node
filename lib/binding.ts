import * as path from "path";

export type ChatMessage = {
  role: string;
  content: string;
};

export type LlamaModelOptions = {
  model: string;
  chat_template?: string;
  embedding?: boolean;
  embd_normalize?: number;
  pooling_type?: "none" | "mean" | "cls" | "last" | "rank";
  n_ctx?: number;
  n_batch?: number;
  n_ubatch?: number;
  n_threads?: number;
  n_gpu_layers?: number;
  flash_attn?: boolean;
  cache_type_k?:
    | "f16"
    | "f32"
    | "q8_0"
    | "q4_0"
    | "q4_1"
    | "iq4_nl"
    | "q5_0"
    | "q5_1";
  cache_type_v?:
    | "f16"
    | "f32"
    | "q8_0"
    | "q4_0"
    | "q4_1"
    | "iq4_nl"
    | "q5_0"
    | "q5_1";
  use_mlock?: boolean;
  use_mmap?: boolean;
  vocab_only?: boolean;
  lora?: string;
  lora_scaled?: number;
  lora_list?: { path: string; scaled: number }[];
};

export type CompletionResponseFormat = {
  type: "text" | "json_object" | "json_schema";
  json_schema?: {
    strict?: boolean;
    schema: object;
  };
  schema?: object; // for json_object type
};

export type LlamaCompletionOptions = {
  messages?: ChatMessage[];
  jinja?: boolean;
  chat_template?: string;
  response_format?: CompletionResponseFormat;
  tools?: object;
  parallel_tool_calls?: boolean;
  tool_choice?: string;
  prompt?: string;
  temperature?: number;
  top_k?: number;
  top_p?: number;
  min_p?: number;
  mirostat?: number;
  mirostat_tau?: number;
  mirostat_eta?: number;
  penalty_last_n?: number;
  penalty_repeat?: number;
  penalty_freq?: number;
  penalty_present?: number;
  typ_p?: number;
  xtc_threshold?: number;
  xtc_probability?: number;
  dry_multiplier?: number;
  dry_base?: number;
  dry_allowed_length?: number;
  dry_penalty_last_n?: number;
  n_predict?: number;
  max_length?: number;
  max_tokens?: number;
  seed?: number;
  stop?: string[];
  grammar?: string;
  grammar_lazy?: boolean;
  grammar_triggers?: { word: string; at_start: boolean }[];
  preserved_tokens?: string[];
};

export type LlamaCompletionResult = {
  text: string;
  tokens_predicted: number;
  tokens_evaluated: number;
  truncated: boolean;
  timings: {
    prompt_n: number;
    prompt_ms: number;
    prompt_per_token_ms: number;
    prompt_per_second: number;
    predicted_n: number;
    predicted_ms: number;
    predicted_per_token_ms: number;
    predicted_per_second: number;
  };
};

export type LlamaCompletionToken = {
  token: string;
};

export type TokenizeResult = {
  tokens: Int32Array;
};

export type EmbeddingResult = {
  embedding: Float32Array;
};

export interface LlamaContext {
  new (options: LlamaModelOptions): LlamaContext;
  getSystemInfo(): string;
  getModelInfo(): object;
  getFormattedChat(
    messages: ChatMessage[],
    chat_template?: string,
    params?: {
      jinja?: boolean;
      response_format?: CompletionResponseFormat;
      tools?: object;
      parallel_tool_calls?: object;
      tool_choice?: string;
    }
  ): object | string;
  completion(
    options: LlamaCompletionOptions,
    callback?: (token: LlamaCompletionToken) => void
  ): Promise<LlamaCompletionResult>;
  stopCompletion(): void;
  tokenize(text: string): Promise<TokenizeResult>;
  detokenize(tokens: number[]): Promise<string>;
  embedding(text: string): Promise<EmbeddingResult>;
  saveSession(path: string): Promise<void>;
  loadSession(path: string): Promise<void>;
  release(): Promise<void>;
  applyLoraAdapters(adapters: { path: string; scaled: number }[]): void;
  removeLoraAdapters(adapters: { path: string }[]): void;
  getLoadedLoraAdapters(): { path: string; scaled: number }[];
  // static
  loadModelInfo(path: string, skip: string[]): Promise<Object>;
}

export interface Module {
  LlamaContext: LlamaContext;
}

export type LibVariant = "default" | "vulkan" | "cuda";

const setupEnv = (variant?: string) => {
  const postfix = variant ? `-${variant}` : "";
  const binPath = path.resolve(
    __dirname,
    `../bin/${process.platform}${postfix}/${process.arch}/`
  );
  const systemPathEnv = process.env.PATH ?? process.env.Path ?? "";
  if (!systemPathEnv.includes(binPath)) {
    if (process.platform === "win32") {
      process.env.Path = `${binPath};${systemPathEnv}`;
    } else {
      process.env.PATH = `${binPath}:${systemPathEnv}`;
    }
  }
};

export const loadModule = async (variant?: LibVariant): Promise<Module> => {
  try {
    if (variant && variant !== "default") {
      setupEnv(variant);
      return (await import(
        `../bin/${process.platform}-${variant}/${process.arch}/llama-node.node`
      )) as Module;
    }
  } catch {} // ignore errors and try the common path
  setupEnv();
  return (await import(
    `../bin/${process.platform}/${process.arch}/llama-node.node`
  )) as Module;
};
