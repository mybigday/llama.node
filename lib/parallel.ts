// Parallel decoding API implementation for llama.node
import type {
  LlamaContext,
  LlamaCompletionOptions,
  LlamaCompletionToken,
  RerankParams,
} from './binding'

export class LlamaParallelAPI {
  private context: LlamaContext
  private enabled: boolean = false
  private pendingRequests = new Map<
    number,
    {
      resolve: (value: any) => void
      reject: (reason?: any) => void
    }
  >()

  constructor(context: LlamaContext) {
    this.context = context
  }

  /**
   * Enable parallel decoding mode
   * @param config Configuration for parallel mode
   * @returns boolean indicating if successful
   */
  async enable(config?: {
    n_parallel?: number
    n_batch?: number
  }): Promise<boolean> {
    const defaultConfig = { n_parallel: 2, n_batch: 512 }
    const result = this.context.enableParallelMode({
      ...defaultConfig,
      ...config,
    })
    this.enabled = result
    return result
  }

  /**
   * Disable parallel decoding mode
   */
  disable(): void {
    this.context.disableParallelMode()
    this.enabled = false
  }

  /**
   * Configure parallel decoding mode (enables if not already enabled)
   * @param config Configuration for parallel mode
   * @returns boolean indicating if successful
   */
  async configure(config: {
    n_parallel?: number
    n_batch?: number
  }): Promise<boolean> {
    return this.enable(config)
  }

  /**
   * Queue a completion request for parallel processing
   * @param options Completion options
   * @param onToken Optional callback for each token
   * @returns Object with requestId, promise for result, and stop function
   */
  async completion(
    options: LlamaCompletionOptions,
    onToken?: (requestId: number, data: LlamaCompletionToken) => void,
  ): Promise<{
    requestId: number
    promise: Promise<any>
    stop: () => void
  }> {
    if (!this.enabled) {
      throw new Error('Parallel mode is not enabled. Call enable() first.')
    }

    const tokenCallback = onToken
      ? (error: any, result: any) => {
          if (error) {
            console.error('Token callback error:', error)
            // Handle completion error
            const pendingReq = this.pendingRequests.get(result?.requestId)
            if (pendingReq) {
              pendingReq.reject(error)
              this.pendingRequests.delete(result?.requestId)
            }
            return
          }
          // Check if this is a token callback or final result
          if (result) {
            if (result.token !== undefined) {
              // This is a token callback
              onToken(result.requestId, result)
            } else if (
              result.text !== undefined ||
              result.content !== undefined
            ) {
              // This is the final result
              const pendingReq = this.pendingRequests.get(result.requestId)
              if (pendingReq) {
                pendingReq.resolve(result)
                this.pendingRequests.delete(result.requestId)
              }
            }
          }
        }
      : undefined

    // Queue the completion immediately (this is synchronous!)
    const { requestId } = this.context.queueCompletion(
      options,
      tokenCallback ||
        ((error, result) => {
          if (error) {
            const pendingReq = this.pendingRequests.get(result?.requestId)
            if (pendingReq) {
              pendingReq.reject(error)
              this.pendingRequests.delete(result?.requestId)
            }
          } else if (
            result &&
            (result.text !== undefined || result.content !== undefined)
          ) {
            // Final result for non-streaming
            const pendingReq = this.pendingRequests.get(result.requestId)
            if (pendingReq) {
              pendingReq.resolve(result)
              this.pendingRequests.delete(result.requestId)
            }
          }
        }),
    )

    // Create promise for final result
    const promise = new Promise((resolveResult, rejectResult) => {
      this.pendingRequests.set(requestId, {
        resolve: resolveResult,
        reject: rejectResult,
      })
    })

    // Create stop function
    const stop = () => {
      this.context.cancelRequest(requestId)
      const pendingReq = this.pendingRequests.get(requestId)
      if (pendingReq) {
        pendingReq.reject(new Error('Request cancelled'))
        this.pendingRequests.delete(requestId)
      }
    }

    // Return immediately without wrapping in a Promise
    return {
      requestId,
      promise,
      stop,
    }
  }

  /**
   * Queue an embedding request for parallel processing
   * @param text Text to embed
   * @param params Optional embedding parameters
   * @returns Object with requestId and promise for result
   */
  async embedding(
    text: string,
    params?: { embd_normalize?: number },
  ): Promise<{
    requestId: number
    promise: Promise<{ embedding: number[] }>
  }> {
    if (!this.enabled) {
      throw new Error('Parallel mode is not enabled. Call enable() first.')
    }

    // Create promise for result
    let resolveResult: (value: any) => void
    let rejectResult: (reason?: any) => void

    const promise = new Promise<{ embedding: number[] }>((res, rej) => {
      resolveResult = res
      rejectResult = rej
    })

    // Queue the embedding immediately (this is synchronous!)
    const { requestId } = this.context.queueEmbedding(
      text,
      params,
      (error, result) => {
        if (error) {
          rejectResult(error)
        } else {
          resolveResult(result)
        }
      },
    )

    // Return immediately without wrapping in a Promise
    return {
      requestId,
      promise,
    }
  }

  /**
   * Queue a rerank request for parallel processing
   * @param query Query text
   * @param documents Documents to rank
   * @param params Optional rerank parameters
   * @returns Object with requestId and promise for results
   */
  async rerank(
    query: string,
    documents: string[],
    params?: RerankParams,
  ): Promise<{
    requestId: number
    promise: Promise<Array<{ score: number; index: number; document: string }>>
  }> {
    if (!this.enabled) {
      throw new Error('Parallel mode is not enabled. Call enable() first.')
    }

    // Create promise for result
    let resolveResult: (value: any) => void
    let rejectResult: (reason?: any) => void

    const promise = new Promise<
      Array<{ score: number; index: number; document: string }>
    >((res, rej) => {
      resolveResult = res
      rejectResult = rej
    })

    // Queue the rerank immediately (this is synchronous!)
    const { requestId } = this.context.queueRerank(
      query,
      documents,
      params,
      (error, result) => {
        if (error) {
          rejectResult(error)
        } else {
          // Add document text to results and sort by score
          const enrichedResults = result.results
            .map((r: any) => ({
              ...r,
              document: documents[r.index],
            }))
            .sort((a: any, b: any) => b.score - a.score)
          resolveResult(enrichedResults)
        }
      },
    )

    // Return immediately without wrapping in a Promise
    return {
      requestId,
      promise,
    }
  }

  /**
   * Check if parallel mode is enabled
   */
  isEnabled(): boolean {
    return this.enabled
  }
}
