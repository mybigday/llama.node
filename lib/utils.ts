
import type {
  ChatMessage,
} from './binding'

export const MTMD_DEFAULT_MEDIA_MARKER = '<__media__>'

export const formatMediaChat = (messages: ChatMessage[] | undefined): {
  messages: ChatMessage[] | undefined
  has_media: boolean
  media_paths?: string[]
} => {
  if (!messages)
    return {
      messages,
      has_media: false,
    }
  const mediaPaths: string[] = []
  return {
    messages: messages.map((msg) => {
      if (Array.isArray(msg.content)) {
        const content = msg.content.map((part) => {
          // Handle multimodal content
          if (part.type === 'image_url') {
            let path = part.image_url?.url || ''
            mediaPaths.push(path)
            return {
              type: 'text',
              text: MTMD_DEFAULT_MEDIA_MARKER,
            }
          } else if (part.type === 'input_audio') {
            const { input_audio: audio } = part
            if (!audio) throw new Error('input_audio is required')

            const { format } = audio
            if (format != 'wav' && format != 'mp3') {
              throw new Error(`Unsupported audio format: ${format}`)
            }
            if (audio.url) {
              const path = audio.url.replace(/file:\/\//, '')
              mediaPaths.push(path)
            } else if (audio.data) {
              mediaPaths.push(audio.data)
            }
            return {
              type: 'text',
              text: MTMD_DEFAULT_MEDIA_MARKER,
            }
          }
          return part
        })

        return {
          ...msg,
          content,
        }
      }
      return msg
    }),
    has_media: mediaPaths.length > 0,
    media_paths: mediaPaths,
  }
}