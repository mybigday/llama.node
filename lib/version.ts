export const BUILD_NUMBER = '6096';
export const BUILD_COMMIT = 'fd1234cb';
