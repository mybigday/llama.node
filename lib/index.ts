import { loadModule, LlamaModelOptions } from './binding'
import type {
  Module,
  LlamaContext,
  LibVariant,
  ChatMessage,
  LlamaCompletionOptions,
  LlamaCompletionToken,
  LlamaCompletionResult,
  TokenizeResult,
  EmbeddingResult,
  RerankParams,
  RerankResult,
  CompletionResponseFormat,
  ModelInfo,
  JinjaFormattedChatResult,
  Tool,
  GGUFModelInfo,
} from './binding'
import { BUILD_NUMBER, BUILD_COMMIT } from './version'
import { LlamaParallelAPI } from './parallel'

export * from './binding'
export { LlamaParallelAPI }

export const MTMD_DEFAULT_MEDIA_MARKER = '<__media__>'

export interface LlamaModelOptionsExtended extends LlamaModelOptions {
  lib_variant?: LibVariant
}

const mods: { [key: string]: Module } = {}

const logListeners: Array<(level: string, text: string) => void> = []

const logCallback = (level: string, text: string) => {
  logListeners.forEach((listener) => listener(level, text))
}

let logEnabled = false

const refreshNativeLogSetup = () => {
  Object.entries(mods).forEach(([, mod]) => {
    mod.LlamaContext.toggleNativeLog(logEnabled, logCallback)
  })
}

export const toggleNativeLog = async (enable: boolean) => {
  logEnabled = enable
  refreshNativeLogSetup()
}

export function addNativeLogListener(
  listener: (level: string, text: string) => void,
): { remove: () => void } {
  logListeners.push(listener)
  return {
    remove: () => {
      logListeners.splice(logListeners.indexOf(listener), 1)
    },
  }
}

const getJsonSchema = (responseFormat?: CompletionResponseFormat) => {
  if (responseFormat?.type === 'json_schema') {
    return responseFormat.json_schema?.schema
  }
  if (responseFormat?.type === 'json_object') {
    return responseFormat.schema || {}
  }
  return null
}

export type FormattedChatResult = {
  type: 'jinja' | 'llama-chat'
  prompt: string
  has_media: boolean
  media_paths?: Array<string>
}

class LlamaContextWrapper {
  ctx: LlamaContext
  parallel: LlamaParallelAPI

  constructor(nativeCtx: LlamaContext) {
    this.ctx = nativeCtx
    this.parallel = new LlamaParallelAPI(nativeCtx)
  }

  getSystemInfo(): string {
    return this.ctx.getSystemInfo()
  }

  getModelInfo(): ModelInfo {
    return this.ctx.getModelInfo()
  }

  isJinjaSupported(): boolean {
    const { minja } = this.ctx.getModelInfo().chatTemplates
    return !!minja?.toolUse || !!minja?.default
  }

  isLlamaChatSupported(): boolean {
    return !!this.ctx.getModelInfo().chatTemplates.llamaChat
  }

  _formatMediaChat(messages: ChatMessage[] | undefined): {
    messages: ChatMessage[] | undefined
    has_media: boolean
    media_paths?: string[]
  } {
    if (!messages)
      return {
        messages,
        has_media: false,
      }
    const mediaPaths: string[] = []
    return {
      messages: messages.map((msg) => {
        if (Array.isArray(msg.content)) {
          const content = msg.content.map((part) => {
            // Handle multimodal content
            if (part.type === 'image_url') {
              let path = part.image_url?.url || ''
              mediaPaths.push(path)
              return {
                type: 'text',
                text: MTMD_DEFAULT_MEDIA_MARKER,
              }
            } else if (part.type === 'input_audio') {
              const { input_audio: audio } = part
              if (!audio) throw new Error('input_audio is required')

              const { format } = audio
              if (format != 'wav' && format != 'mp3') {
                throw new Error(`Unsupported audio format: ${format}`)
              }
              if (audio.url) {
                const path = audio.url.replace(/file:\/\//, '')
                mediaPaths.push(path)
              } else if (audio.data) {
                mediaPaths.push(audio.data)
              }
              return {
                type: 'text',
                text: MTMD_DEFAULT_MEDIA_MARKER,
              }
            }
            return part
          })

          return {
            ...msg,
            content,
          }
        }
        return msg
      }),
      has_media: mediaPaths.length > 0,
      media_paths: mediaPaths,
    }
  }

  getFormattedChat(
    messages: ChatMessage[],
    template?: string,
    params?: {
      jinja?: boolean
      response_format?: CompletionResponseFormat
      tools?: Tool[]
      parallel_tool_calls?: boolean
      tool_choice?: string
      enable_thinking?: boolean
      add_generation_prompt?: boolean
      now?: string | number
      chat_template_kwargs?: Record<string, string>
    },
  ): FormattedChatResult {
    const {
      messages: chat,
      has_media,
      media_paths,
    } = this._formatMediaChat(messages)

    const useJinja = this.isJinjaSupported() && params?.jinja
    let tmpl
    if (template) tmpl = template // Force replace if provided

    const result = this.ctx.getFormattedChat(chat!, tmpl, {
      jinja: useJinja,
      response_format: params?.response_format,
      tools: params?.tools,
      parallel_tool_calls: params?.parallel_tool_calls,
      tool_choice: params?.tool_choice,
      enable_thinking: params?.enable_thinking ?? true,
      add_generation_prompt: params?.add_generation_prompt,
      now: params?.now,
      chat_template_kwargs: params?.chat_template_kwargs
        ? Object.entries(params.chat_template_kwargs).reduce(
            (acc, [key, value]) => {
              acc[key] = JSON.stringify(value) // Each value is a stringified JSON object
              return acc
            },
            {} as Record<string, any>,
          )
        : undefined,
    })

    if (!useJinja) {
      return {
        type: 'llama-chat',
        prompt: result as string,
        has_media,
        media_paths,
      }
    }
    const jinjaResult = result as JinjaFormattedChatResult
    return {
      type: 'jinja',
      has_media,
      media_paths,
      ...jinjaResult,
    }
  }

  completion(
    options: LlamaCompletionOptions,
    callback?: (token: LlamaCompletionToken) => void,
  ): Promise<LlamaCompletionResult> {
    const { messages, media_paths = options.media_paths } =
      this._formatMediaChat(options.messages)
    return this.ctx.completion(
      {
        ...options,
        messages,
        media_paths: options.media_paths || media_paths,
      },
      callback || (() => {}),
    )
  }

  stopCompletion(): void {
    return this.ctx.stopCompletion()
  }

  tokenize(
    text: string,
    { media_paths }: { media_paths?: string[] } = {},
  ): Promise<TokenizeResult> {
    return this.ctx.tokenize(text, media_paths)
  }

  detokenize(tokens: number[]): Promise<string> {
    return this.ctx.detokenize(tokens)
  }

  embedding(text: string, params?: { embd_normalize?: number }): Promise<EmbeddingResult> {
    return this.ctx.embedding(text, params)
  }

  rerank(
    query: string,
    documents: string[],
    params?: RerankParams,
  ): Promise<Array<RerankResult & { document: string }>> {
    return this.ctx
      .rerank(query, documents, params)
      .then((results: RerankResult[]) => {
        // Sort by score descending and add document text for convenience
        return results
          .map((result: RerankResult) => ({
            ...result,
            document: documents[result.index],
          }))
          .sort(
            (
              a: RerankResult & { document: string },
              b: RerankResult & { document: string },
            ) => b.score - a.score,
          )
      })
  }

  saveSession(path: string): Promise<void> {
    return this.ctx.saveSession(path)
  }

  loadSession(path: string): Promise<void> {
    return this.ctx.loadSession(path)
  }

  release(): Promise<void> {
    return this.ctx.release()
  }

  applyLoraAdapters(adapters: { path: string; scaled: number }[]): void {
    return this.ctx.applyLoraAdapters(adapters)
  }

  removeLoraAdapters(): void {
    this.ctx.removeLoraAdapters()
  }

  getLoadedLoraAdapters(): { path: string; scaled: number }[] {
    return this.ctx.getLoadedLoraAdapters()
  }

  initMultimodal(options: { path: string; use_gpu?: boolean }): boolean {
    return this.ctx.initMultimodal(options)
  }

  isMultimodalEnabled(): boolean {
    return this.ctx.isMultimodalEnabled()
  }

  releaseMultimodal(): void {
    this.ctx.releaseMultimodal()
  }

  getMultimodalSupport(): {
    vision: boolean
    audio: boolean
  } {
    return this.ctx.getMultimodalSupport()
  }

  initVocoder(options: { path: string; n_batch?: number }): boolean {
    return this.ctx.initVocoder(options)
  }

  releaseVocoder(): void {
    this.ctx.releaseVocoder()
  }

  isVocoderEnabled(): boolean {
    return this.ctx.isVocoderEnabled()
  }

  getFormattedAudioCompletion(
    speaker: string | null,
    text: string,
  ): {
    prompt: string
    grammar?: string
  } {
    return this.ctx.getFormattedAudioCompletion(speaker, text)
  }

  getAudioCompletionGuideTokens(text: string): Int32Array {
    return this.ctx.getAudioCompletionGuideTokens(text)
  }

  decodeAudioTokens(tokens: number[] | Int32Array): Promise<Float32Array> {
    return this.ctx.decodeAudioTokens(tokens)
  }
}

export const loadModel = async (
  options: LlamaModelOptionsExtended,
): Promise<LlamaContextWrapper> => {
  const variant = options.lib_variant ?? 'default'
  mods[variant] ??= await loadModule(options.lib_variant)
  refreshNativeLogSetup()

  const nativeCtx = new mods[variant].LlamaContext(options)
  return new LlamaContextWrapper(nativeCtx)
}

export const initLlama = loadModule

const modelInfoSkip = [
  // Large fields
  'tokenizer.ggml.tokens',
  'tokenizer.ggml.token_type',
  'tokenizer.ggml.merges',
  'tokenizer.ggml.scores',
]

export const loadLlamaModelInfo = async (
  path: string,
): Promise<GGUFModelInfo> => {
  const variant = 'default'
  mods[variant] ??= await loadModule(variant)
  refreshNativeLogSetup()
  return mods[variant].LlamaContext.loadModelInfo(path, modelInfoSkip)
}

export const BuildInfo = {
  number: BUILD_NUMBER,
  commit: BUILD_COMMIT,
}
