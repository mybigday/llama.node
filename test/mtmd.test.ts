import path from 'path'
import { loadModel } from '../lib'

jest.setTimeout(60000)

test('multimodal with images', async () => {
  const model = await loadModel({
    model: path.resolve(__dirname, './SmolVLM-256M-Instruct-Q8_0.gguf'),
    // model: path.resolve(__dirname, './ggml-org_gemma-3-4b-it-GGUF_gemma-3-4b-it-Q4_K_M.gguf'),
    n_gpu_layers: 0,
    n_ctx: 512,
  })

  // Initialize multimodal support
  const mmproj_path = path.resolve(
    __dirname,
    './mmproj-SmolVLM-256M-Instruct-Q8_0.gguf',
  )
  // const mmproj_path = path.resolve(__dirname, './ggml-org_gemma-3-4b-it-GGUF_mmproj-model-f16.gguf')
  await model.initMultimodal({ path: mmproj_path, use_gpu: false })

  // Test with multiple images
  const result = await model.completion({
    messages: [
      {
        role: 'system',
        content: 'You are a helpful assistant that can answer questions and help with tasks.',
      },
      {
        role: 'user',
        content: [
          {
            type: 'text', 
            text: 'What is the content of this image?',
          },
          {
            type: 'image_url',
            image_url: {
              url: path.resolve(__dirname, './test-1.jpeg'),
            },
          },
        ],
      },
    ],
    temperature: 0,
    n_predict: 100,
    seed: 0,
  })

  expect(result.text.length).toBeGreaterThan(0)
  expect(result.content).toBe(result.text)
  expect(result).toMatchObject({
    tokens_predicted: expect.any(Number),
    tokens_evaluated: expect.any(Number),
    text: expect.stringMatching(/ A newspaper with/),
  })

  console.log(result)

  expect(await model.isMultimodalEnabled()).toBe(true)
  await model.releaseMultimodal()
  expect(await model.isMultimodalEnabled()).toBe(false)

  await model.release()
})
