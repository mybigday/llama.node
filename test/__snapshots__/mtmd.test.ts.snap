// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multimodal with images 1`] = `
{
  "bitmap_hashes": [
    "848984200801287295",
  ],
  "chunk_pos": [
    0,
    30,
    94,
  ],
  "chunk_pos_images": [
    30,
  ],
  "has_image": true,
  "tokens": Int32Array [
    1,
    2683,
    359,
    253,
    5356,
    11173,
    338,
    416,
    2988,
    2029,
    284,
    724,
    351,
    4381,
    30,
    198,
    198,
    11126,
    42,
    1812,
    314,
    260,
    2627,
    282,
    451,
    2443,
    47,
    198,
    49189,
    49152,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    49189,
    49279,
    198,
    9519,
    9531,
    42,
  ],
}
`;

exports[`multimodal with images 2`] = `
{
  "has_image": true,
  "image_paths": [
    "<img-path>",
  ],
  "prompt": "<|im_start|>You are a helpful assistant that can answer questions and help with tasks.

User: What is the content of this image?
<__image__><end_of_utterance>
Assistant:",
  "type": "llama-chat",
}
`;
